src/or/main.c

```diff
@@ -1348,37 +1348,54 @@ CALLBACK(write_stats_file);
 #undef CALLBACK
 
 /* Now we declare an array of periodic_event_item_t for each periodic event */
-#define CALLBACK(name) PERIODIC_EVENT(name)
+#define CALLBACK(name, r) PERIODIC_EVENT(name, r)
 
 static periodic_event_item_t periodic_events[] = {
-  CALLBACK(add_entropy),
-  CALLBACK(check_authority_cert),
-  CALLBACK(check_canonical_channels),
-  CALLBACK(check_descriptor),
-  CALLBACK(check_dns_honesty),
-  CALLBACK(check_ed_keys),
-  CALLBACK(check_expired_networkstatus),
-  CALLBACK(check_for_reachability_bw),
-  CALLBACK(check_onion_keys_expiry_time),
-  CALLBACK(clean_caches),
-  CALLBACK(clean_consdiffmgr),
-  CALLBACK(downrate_stability),
-  CALLBACK(expire_old_ciruits_serverside),
-  CALLBACK(fetch_networkstatus),
-  CALLBACK(heartbeat),
-  CALLBACK(hs_service),
-  CALLBACK(launch_descriptor_fetches),
-  CALLBACK(launch_reachability_tests),
-  CALLBACK(record_bridge_stats),
-  CALLBACK(rend_cache_failure_clean),
-  CALLBACK(reset_padding_counts),
-  CALLBACK(retry_dns),
-  CALLBACK(retry_listeners),
-  CALLBACK(rotate_onion_key),
-  CALLBACK(rotate_x509_certificate),
-  CALLBACK(save_stability),
-  CALLBACK(write_bridge_ns),
-  CALLBACK(write_stats_file),
+  /* Everyone needs to run those. */
+  CALLBACK(add_entropy, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(check_expired_networkstatus, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(clean_caches, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(fetch_networkstatus, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(heartbeat, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(launch_descriptor_fetches, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(reset_padding_counts, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(retry_listeners, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(rotate_x509_certificate, PERIODIC_EVENT_ROLE_ALL),
+  CALLBACK(write_stats_file, PERIODIC_EVENT_ROLE_ALL),
+
+  /* Routers (bridge and relay) only. */
+  CALLBACK(check_descriptor, PERIODIC_EVENT_ROLE_ROUTER),
+  CALLBACK(check_ed_keys, PERIODIC_EVENT_ROLE_ROUTER),
+  CALLBACK(check_for_reachability_bw, PERIODIC_EVENT_ROLE_ROUTER),
+  CALLBACK(check_onion_keys_expiry_time, PERIODIC_EVENT_ROLE_ROUTER),
+  CALLBACK(clean_consdiffmgr, PERIODIC_EVENT_ROLE_ROUTER),
```

Comment:
I think this needs a separate "DirServer" role, unless it is going to be "Everyone".  We have a bug recently where we assumed that all DirServers were relays of some kind, but that doesn't seem to be true.. See trac:23693

Reply:
Again as above, `server_mode()` is used meaning bridge go in that path. And because all relays are `DirServer`, it seems OK to assign the Router role where a bridge also will serve consensus to their clients thus need to clean the cache?

Reply:
Per IRC discussion, we need a DirServer role since `DirCache 0` will make this callback pointless for a relay.

Reply:
I opened #25900 for this